### PR TITLE
FEAT: implement quantum problem set filter

### LINKED ISSUE #297: Give an example for the usage of the `filter_quantum_number_problem_set`-function in a notebook
In #287 an auxilliary function `filter_quantum_number_problem_set` arose which may be useful to users and thus should be explained in an example in the `visualize`-notebook. This one is suited, as it already discusses the `QuantumNumberProblemSet`s in contrast to the particle based `ProblemSet`s.

### PR DESCRIPTION
**Pull request by grayson-helmholz**
As `CSPSolver.find_solutions()` just takes a general `QNProblemSet` it should be possible 
to  reduce the set of constraints (i.e. conservation rules) the solver takes into account for 
finding the solutions of a decay. 
A possible way to implement this is to use the `StateTransitionManager` to generate the 
full set of rules for a certain decay and just filter it to contain only the desired 
constraints before passing it to `CSPSolver`.

Closes #297 

> [!TIP]
> See preview of the modified documentation [here](https://qrules--287.org.readthedocs.build/287/usage/visualize.html#filtering-quantum-number-problem-sets).

### CODE REVIEW

**Review by redeboer — CHANGES_REQUESTED** [review]
Ported remaining review comments from https://github.com/ComPWA/qrules/pull/278

**Review by redeboer — CHANGES_REQUESTED** [review]
Nice that it's working 🙌 

(After offline discussion):
Let's expose this `filter_quantum_number_problem_set()` function and write an example with visualization in the "Conservation rules" notebook. If that turns out to look concise, illustrative, and useable enough, we can keep this new function in the API.

File: tests/unit/test_solving.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+        edge_rules={spin_validity},
+        node_rules={
+            spin_magnitude_conservation,
+            parity_conservation,
+            c_parity_conservation,
+        },
+        edge_properties_and_domains={
+            EdgeQuantumNumbers.pid,  # had to be added for c_parity_conservation to work
+            EdgeQuantumNumbers.spin_magnitude,
+            # EdgeQuantumNumbers.spin_projection,  # can be left out to reduce the number of solutions
```
[the remark this answers is not in the document]
Comment (grayson-helmholz) [reply tests/unit/test_solving.py:44]: implemented

File: docs/usage/conservation.ipynb [hunk 2]
```diff
@@ -241,9 +241,6 @@
    "cell_type": "code",
    "execution_count": null,
    "metadata": {
```
Comment (redeboer) [inline docs/usage/conservation.ipynb:243]: Could you undo the changes to the notebooks that you did not intend to edit in this PR?

**Review by redeboer — CHANGES_REQUESTED** [review]
Good job! The new section in the documentation is particularly good 👌 
Some final smaller changes are suggested below.

File: tests/unit/test_solving.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        final_state=["gamma", "eta", "eta"],
+        formalism="helicity",
+    )
+    problem_sets = stm.create_problem_sets()
+    qn_problem_sets = [
+        p.to_qn_problem_set()
+        for strength in sorted(problem_sets)
+        for p in problem_sets[strength]
+    ]
+    return qn_problem_sets[0]
```
[the remark this answers is not in the document]
Comment (grayson-helmholz) [reply tests/unit/test_solving.py:90]: Yes, should have had this built-in in mind.

File: src/qrules/solving.py [hunk 4]
```diff
@@ -101,6 +101,86 @@ def topology(self) -> Topology:
         return self.initial_facts.topology
 
 
+def filter_quantum_number_problem_set(
+    quantum_number_problem_set: QNProblemSet,
+    edge_rules: set[GraphElementRule],
+    node_rules: set[Rule],
+    edge_properties_and_domains: Iterable[EdgeQuantumNumberTypes],
+    node_properties_and_domains: Iterable[NodeQuantumNumberTypes],
```
[the remark this answers is not in the document]
Comment (grayson-helmholz) [reply src/qrules/solving.py:109]: When this function was originally written to test the behaviour in the `merge_rules&priorities`-branch and used `attrs.evolve`, this was supposed to signal during development that it also copied the domains and not only the (fused) rules and priorities.
Will be renamed as it now constructs whole new Settings anyway.

File: docs/usage/visualize.ipynb [hunk 5]
```diff
@@ -315,6 +326,96 @@
     "graphviz.Source(dot)"
    ]
   },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Filtering Quantum Number Problem Sets\n",
```
[the remark this answers is not in the document]
Comment (grayson-helmholz) [reply docs/usage/visualize.ipynb:333]: Makes sense, did not think about that.

File: docs/usage/visualize.ipynb [hunk 6]
```diff
@@ -315,6 +326,96 @@
...
+     "source_hidden": true
+    },
+    "tags": [
+     "hide-cell"
+    ]
+   },
+   "outputs": [],
+   "source": [
+    "dot = qrules.io.asdot(filtered_qn_result, render_node=True)\n",
+    "graphviz.Source(dot)"
```
[the remark this answers is not in the document]
Comment (grayson-helmholz) [reply docs/usage/visualize.ipynb:416]: Will post an issue about that

[the remark this answers is not in the document]
Comment (grayson-helmholz) [reply docs/usage/visualize.ipynb:416]: Is now mentioned in #301

[the remark this answers is not in the document]
Comment (redeboer) [reply docs/usage/visualize.ipynb:416]: It has to do with the quantum number sets from which the `CSPSolver` gets its info, specifically here:
```python
solver = CSPSolver([
    qrules.system_control.create_edge_properties(part)
    for part in qrules.particle.load_pdg()
])
```
So you probably want to filter the required properties from `qrules.system_control.create_edge_properties(part)`.

**Review by redeboer — APPROVED** [review]
Good job! 🥇

**Pull request merged**